torchvision/prototype/datasets/_builtin/sintel.py

```diff
@@ -0,0 +1,182 @@
+        return split in path.parent.parent.parent.name
+
+    def _filter_images(self, data: Tuple[str, Any], *, pass_name: str) -> bool:
+        path = pathlib.Path(data[0])
+        return pass_name in str(path.parent) and path.suffix == ".png"
+
+    def _classify_archive(self, data: Tuple[str, Any], *, pass_name: str) -> Optional[int]:
+        path = pathlib.Path(data[0])
+        suffix = path.suffix
+        if suffix == ".flo":
+            return 0
+        elif pass_name == path.parent.parent.name and suffix == ".png":
+            return 1
+        else:
+            return None
+
+    def _read_flo(self, file: io.IOBase) -> torch.Tensor:
+        magic = file.read(4)
+        if magic != b"PIEH":
+            raise ValueError("Magic number incorrect. Invalid .flo file")
+        w = int.from_bytes(file.read(4), "little")
+        h = int.from_bytes(file.read(4), "little")
+        data = file.read(2 * w * h * 4)
+        data_arr = np.frombuffer(data, dtype=np.float32)
+        # Creating a copy of the underlying array, to avoid UserWarning: "The given NumPy array
+        #     is not writeable, and PyTorch does not support non-writeable tensors."
+        return torch.from_numpy(np.copy(data_arr.reshape(h, w, 2).transpose(2, 0, 1)))
+
+    def _flows_key(self, data: Tuple[str, Any]) -> Tuple[str, int]:
+        path = pathlib.Path(data[0])
+        category = path.parent.name
+        idx = int(self._FILE_NAME_PATTERN.match(path.name).group("idx"))  # type: ignore[union-attr]
+        return category, idx
+
+    def _images_key(self, data: Tuple[Tuple[str, Any], Tuple[str, Any]]) -> Tuple[str, int]:
+        return self._flows_key(data[0])
+
+    def _collate_and_decode_sample(
+        self,
+        data: Tuple[Tuple[str, io.IOBase], Any],
+        *,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+        config: DatasetConfig,
+    ) -> Dict[str, Any]:
+        # flo, images = data
+        # img1, img2 = images
+
+        if config.split == "train":
+            flo, images = data
+            img1, img2 = images
+            flow_arr = self._read_flo(flo[1])
+        else:
+            # When split is `test`
+            img1, img2 = data
+
+        path1, buffer1 = img1
+        path2, buffer2 = img2
+
+        return dict(
+            image1=(path1, decoder(buffer1)) if decoder else (path1, buffer1),
+            image2=(path2, decoder(buffer2)) if decoder else (path2, buffer2),
+            flow=(flo[0], flow_arr) if config.split == "train" else ("", None),
+        )
+
+    def _make_datapipe(
+        self,
+        resource_dps: List[IterDataPipe],
+        *,
+        config: DatasetConfig,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> IterDataPipe[Dict[str, Any]]:
+        dp = resource_dps[0]
+        archive_dp = ZipArchiveReader(dp)
+
+        curr_split = Filter(archive_dp, self._filter_split, fn_kwargs=dict(split=config.split))
```

Comment:
Can't we have a second `Filter` below this one that filters for the `pass_name`? That way we don't need to handle it twice in `_classify_archive` and in `_filter_images`.

Reply:
I gave it a thought before, but I feel the current implementation saves an extra pass.

Either `_classify_archive` or `_filter_images` function is called depending on `config.split`, but if we have an extra filter below the `curr_split = Filter(..., self._filter_split, ...)` - then that's one pass, and then there's another in the `Demultiplexer` (which uses `_classify_archive`) or in the `Filter` (which uses `_filter_images`). So in total, there are 2 passes through the path name. Not a big difference, I agree, but IMO - the current implementation is okay, as either of them is called depending on the split.

If you feel otherwise, please let me know. Thanks!

Reply:
Not sure I understand. If we have `Filter` on the `pass_name` before we get to the `Demultiplexer`, we don't need to do it inside `_classify_archive`, right? If that is the case, the total number of comparisons stays the same, since `_classify_archive` would only split on the suffix.

Reply:
You are right. Should be fixed now. Sorry for the misunderstanding, and thanks for the suggestion, Philip! :)